### PR TITLE
Pure-Python implementations of patch extraction

### PR DESCRIPTION
**Pull request by jabooth**
- Pure-Python implementations of `set_patches` and `extract_patches` from `patches.pyx`
- Fall back to these if native versions are not available:

## Todo
- Unit test
- Check segfaulting that is happening now when calling set_patches for unit tests

### CODE REVIEW

File: menpo/image/pypatches.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+def extract_patches(pixels, points, patch_shape, offsets):
+    o = (np.array(patch_shape) / 2).astype(np.uint)
+    patches = []
+    for point in points:
+        for sample_offset in offsets:
+            p = (point + sample_offset).astype(np.uint)
+            patches.append(pixels[:, p[0]-o[0]:p[0]+o[0],
+                                     p[1]-o[1]:p[1]+o[1]])
+    return np.array(patches).reshape([points.shape[0], len(offsets),
```
Comment (patricksnape) [inline menpo/image/pypatches.py]: Pre-allocating the array would save memory.

File: menpo/image/pypatches.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+                                      pixels.shape[0], *patch_shape])
+
+
+def set_patches(patches, pixels, points, offsets, offset_index):
+    patch_shape = patches.shape[-2:]
+    o = (np.array(patch_shape) / 2).astype(np.uint)
```
Comment (patricksnape) [inline menpo/image/pypatches.py]: Allocating the array as uint would be cleaner

File: menpo/image/pypatches.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
+# Pure-Python implementations of the Cython functions defined in patches.pyx
+import numpy as np
+
+
+def extract_patches(pixels, points, patch_shape, offsets):
+    o = (np.array(patch_shape) / 2).astype(np.uint)
```
Comment (patricksnape) [inline menpo/image/pypatches.py]: Allocating the array as uint would be cleaner

File: menpo/image/pypatches.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+        for sample_offset in offsets:
+            p = (point + sample_offset).astype(np.uint)
+            patches.append(pixels[:, p[0]-o[0]:p[0]+o[0],
+                                     p[1]-o[1]:p[1]+o[1]])
+    return np.array(patches).reshape([points.shape[0], len(offsets),
+                                      pixels.shape[0], *patch_shape])
+
+
+def set_patches(patches, pixels, points, offsets, offset_index):
+    patch_shape = patches.shape[-2:]
```
Comment (patricksnape) [inline menpo/image/pypatches.py]: Some sort of error checking on the size? So it's clear what shape we expect?

File: menpo/image/pypatches.py [hunk 5]
```diff
@@ -0,0 +1,24 @@
+# Pure-Python implementations of the Cython functions defined in patches.pyx
+import numpy as np
+
+
+def extract_patches(pixels, points, patch_shape, offsets):
```
Comment (patricksnape) [inline menpo/image/pypatches.py]: Documentation for these methods

File: menpo/image/pypatches.py [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    o = (np.array(patch_shape) / 2).astype(np.uint)
+    patches = []
+    for point in points:
+        for sample_offset in offsets:
+            p = (point + sample_offset).astype(np.uint)
+            patches.append(pixels[:, p[0]-o[0]:p[0]+o[0],
+                                     p[1]-o[1]:p[1]+o[1]])
+    return np.array(patches).reshape([points.shape[0], len(offsets),
+                                      pixels.shape[0], *patch_shape])
```
Comment (patricksnape) [inline menpo/image/pypatches.py]: This is invalid syntax in Python 2.x

**Comment by jabooth** [comment]
I've switched the Python patches to the default so for this PR we can see if it passes the test suite for being a complete replacement for the existing cython variant.

Right now we have different behaviour when it comes to patches falling off the image bounds - I haven't checked in detail but it looks like our cython implementation checks if patches would invalidate the image bounds and adjusts the sizes accordingly. I'm not sure if that's a good design tbh, (I ask for (16, 16) and silently get (16, 14)? Need to check in detail and might be me, just going off the two unit tests that are failing (see travis shortly).

**Comment by patricksnape** [comment]
This is slower than the Cython implementation but reduces
the maintenance of the package as a whole by removing the need
for C-extension compilation. It also adds the ability to
interpolate patches with strategies other than nearest neighbour
aka slicing. For example, bilinear interpolation can now be
used. Note this also changes how boundaries are handled as
patches are now always returned and filled with a constant
value rather than being truncated.

For reference, on the masked Takeo image - extracting patches
around the landmarks previously took ~300us on my 2013 Macbook Pro
and now takes ~550us - so close to twice as slow.

**Pull request merged**